src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -47,12 +47,58 @@ public class Regex
 
         // *********** } match timeout fields ***********
 
+        protected internal RegexRunnerFactory factory;
 
-        internal Dictionary<Int32, Int32> _caps;            // if captures are sparse, this is the hashtable capnum->index
-        internal Dictionary<String, Int32> _capnames;       // if named captures are used, this maps names->index
+        internal Dictionary<int, int> _caps;            // if captures are sparse, this is the hashtable capnum->index
+        internal Dictionary<string, int> _capnames;     // if named captures are used, this maps names->index
 
-        internal String[] _capslist;                        // if captures are sparse or named captures are used, this is the sorted list of names
-        internal int _capsize;                              // the size of the capture array
+        protected internal String[] capslist;              // if captures are sparse or named captures are used, this is the sorted list of names
+        protected internal int capsize;                    // the size of the capture array
+        protected IDictionary Caps
+        {
+            get
+            {
+                return _caps;
+            }
+            set
+            {
+                if (value is Dictionary<int, int>)
+                {
+                    _caps = (Dictionary<int, int>)value;
+                }
+                else
+                {
+                    _caps = new Dictionary<int, int>();
+                    foreach(DictionaryEntry entry in value)
+                    {
+                        _caps.Add((int)entry.Key, (int)entry.Value);
+                    }
+                }
```

Comment:
foreach is basically ```IEnumerator``` in IL, and there's only an extra unboxing of the value item to DictionaryEntry.